python/ray/train/torch/torch_checkpoint.py

```diff
@@ -21,6 +25,49 @@ class TorchCheckpoint(Checkpoint):
     ``TorchCheckpoint.from_checkpoint(ckpt)``.
     """
 
+    # Special encoding logic to avoid serialization errors with torch.
+    def _encode_data_dict(self, data_dict: dict) -> dict:
+        """Encode data_dict using torch.save."""
+        from torch.nn.parallel import DistributedDataParallel
+
+        for k, v in data_dict.items():
+            if isinstance(v, DistributedDataParallel) and hasattr(v, "module"):
+                data_dict[k] = v.module
+
+        # Convert the checkpoint dict to bytes, so that any GPU tensors that
+        # are in the checkpoint dict can be properly deserialized on the
+        # driver side, even if the driver does not have access to a GPU device.
+        _buffer = io.BytesIO()
+        torch.save(data_dict, _buffer, pickle_module=ray.cloudpickle)
+        return {ENCODED_DATA_KEY: _buffer.getvalue()}
+
+    def _decode_data_dict(self, data_dict: dict) -> dict:
+        """Decode data_dict using torch_load if needed."""
+        if ENCODED_DATA_KEY not in data_dict:
+            return data_dict
+        encoded_data = data_dict[ENCODED_DATA_KEY]
+        _buffer = io.BytesIO(encoded_data)
+        data_dict = torch.load(
+            _buffer,
+            map_location="cpu"
+            # Not using ray.cloudpickle here as it doesn't
+            # define an Unpickler (as it is not necessary).
+        )
+        return data_dict
+
+    def __getstate__(self) -> dict:
+        if self._data_dict:
+            state = self.__dict__.copy()
```

Comment:
are these copies necessary?

Reply:
we don't want to modify the underlying state when serializing (this is a shallow copy so it should be cheap)